cli/tools/test/channel.rs

```diff
@@ -0,0 +1,481 @@
+  /// because we can't do anything.
+  #[must_use = "If this returns false, don't keep reading because we cannot send"]
+  fn send(&mut self, buffer: Vec<u8>) -> bool {
+    if buffer.is_empty() {
+      true
+    } else if self
+      .sender
+      .send((self.id, TestEvent::Output(self.which, buffer)))
+      .is_err()
+    {
+      self.read_opt.take();
+      false
+    } else {
+      true
+    }
+  }
+
+  fn is_alive(&self) -> bool {
+    self.read_opt.is_some()
+  }
+
+  /// Attempt to read from a given stream, pushing all of the data in it into the given
+  /// [`UnboundedSender`] before returning.
+  async fn pipe(&mut self) {
+    let mut buffer = [0_u8; BUFFER_SIZE];
+    let mut buf = ReadBuf::new(&mut buffer);
+    let res = {
+      // No more stream, so just return.
+      let Some(stream) = &mut self.read_opt else {
+        return;
+      };
+      poll_fn(|cx| Pin::new(&mut *stream).poll_read(cx, &mut buf)).await
+    };
+    match res {
+      Ok(_) => {
+        let buf = buf.filled().to_vec();
+        if buf.is_empty() {
+          // The buffer may return empty in EOF conditions and never return an error,
+          // so we need to treat this as EOF
+          self.read_opt.take();
+        } else {
+          // Attempt to send the buffer, marking as not alive if the channel is closed
+          _ = self.send(buf);
+        }
+      }
+      Err(_) => {
+        // Stream errored, so just return and mark this stream as not alive.
+        _ = self.send(buf.filled().to_vec());
+        self.read_opt.take();
+      }
+    }
+  }
+
+  /// Read and "block" until the sync markers have been read.
+  async fn read_until_sync_marker(&mut self) {
+    let Some(file) = &mut self.read_opt else {
+      return;
+    };
+    let mut flush = Vec::with_capacity(BUFFER_SIZE);
+    loop {
+      let mut buffer = [0_u8; BUFFER_SIZE];
+      match file.read(&mut buffer).await {
+        Err(_) | Ok(0) => {
+          // EOF or error, just return. We make no guarantees about unflushed data at shutdown.
+          self.read_opt.take();
+          return;
+        }
+        Ok(read) => {
+          flush.extend(&buffer[0..read]);
+          if flush.ends_with(SYNC_MARKER) {
+            flush.truncate(flush.len() - SYNC_MARKER.len());
+            // Try to send our flushed buffer. If the channel is closed, this stream will
+            // be marked as not alive.
+            _ = self.send(flush);
+            return;
+          }
+        }
+      }
+    }
+  }
+}
+
+/// A factory for creating [`TestEventSender`]s. This factory must be dropped
+/// before the [`TestEventReceiver`] will complete.
+pub struct TestEventSenderFactory {
+  sender: UnboundedSender<(usize, TestEvent)>,
+  worker_id: AtomicUsize,
+}
+
+impl TestEventSenderFactory {
+  /// Create a [`TestEventWorkerSender`], along with a stdout/stderr stream.
+  pub fn worker(&self) -> TestEventWorkerSender {
+    let id = self.worker_id.fetch_add(1, Ordering::AcqRel);
```

Comment:
I'm a bit confused about these IDs - do we create a `TestEventSenderFactory` for each worker, or is it a singleton for whole `deno test` process?

Reply:
You create a `TestEventSenderFactory` when you have a batch of specifiers to test like the `deno test` command does.  For each specifier, you create one worker. In the case of the REPL and Jupyter, we have a separate method to create one single `TestEventWorkerSender` and a receiver.

The worker allows for parallel execution of tests -- each worker event is associated with its worker ID, allowing us to better interleave parallel results. I didn't actually make it that far to better interleave results, but it will be easier now.